### PR TITLE
Gallery 1.0

### PR DESCRIPTION
**Pull request by Davidspilde**

### CODE REVIEW

**Comment by Davidspilde** [comment]
Lykke til

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [inline Assets/Scripts/UI/GalleryMenu.cs:25]: You could've built the list of grid elements from the return values of the Instantiate calls...

File: Assets/Editor/WeaponDrawer.cs
Comment (toberge) [inline Assets/Editor/WeaponDrawer.cs:9]: Please remember to rework this and delete `CreatePropertyGUI` if it still isn't in use.

File: Assets/Scripts/Augment/GunFactory.cs.orig
Comment (toberge) [inline Assets/Scripts/Augment/GunFactory.cs.orig:1]: Slett denne fila

File: Assets/Scripts/UI/GallaryCreator.cs
Comment (toberge) [inline Assets/Scripts/UI/GallaryCreator.cs:1]: This file seems useless (and its name is misspelled)

File: Assets/Scripts/UI/MainMenuController.cs
Comment (toberge) [inline Assets/Scripts/UI/MainMenuController.cs:24]: Why is this one public now?

**Review by toberge — CHANGES_REQUESTED** [review]
Haven't tested it myself yet, but there are some issues that stick out to me already

**Comment by toberge** [comment]
image

- Hat barrel looks weird.
  - We don't need all the usual initialization logic for the gallery guns, so if we create a method for instantiating a non-functional gun, the hat issue might resolve itself.
- There may be a **lot** of tabs eventually.
- Back button is much larger than on the options menu, should be same size as the back button there.
- The gallery replaces the main menu in the current version. We can't merge it like this.
- And then there's the matter of actually storing the guns that players win with or whatever it is we want. The current gallery will explode in size as we add more gun parts, since it simply shows all possible guns :/

File: Assets/Scripts/UI/MainMenuController.cs
Comment (TArchetype) [reply Assets/Scripts/UI/MainMenuController.cs:24]: We have the following code in GalleryMenu.cs: 
`    public void Back(InputAction.CallbackContext ctx)
    {
        mainMenuController.SwitchToMenu(mainMenuController.defaultMenu);
        mainMenuController.SelectControl(mainMenuController.GetComponentInChildren<Button>());
    }`
    
Which simply allows the menu to be closed by pressing a back button. I plan on expanding this to the other menus as well.

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (TArchetype) [reply Assets/Scripts/UI/GalleryMenu.cs:198]: This collection of structs does not impact performance in any meaningful way and is also only a placeholder function before we implement ways of storing gun statistics and part performance as well as legendary weapons. So I feel this can stay.

File: Assets/Scripts/UI/FlexibleGridLayout.cs
Comment (TArchetype) [reply Assets/Scripts/UI/FlexibleGridLayout.cs:97]: No idea, removed.

File: Assets/Scripts/DataService.cs
Comment (TArchetype) [reply Assets/Scripts/DataService.cs:1]: We might implement statistics through some sort of json files which could later be updated and posted on the Hackerspace website. As the idea never really came to fruition I've deleted the relevant files until we consider implementing it in another branch.

File: Assets/Scripts/Augment/GunFactory.cs
Comment (TArchetype) [reply Assets/Scripts/Augment/GunFactory.cs:24]: The overload has been removed and GallerMenu.cs now sets the parent and scale itself instead.

File: Assets/Scripts/UI/MainMenuController.cs
Comment (Fueredoriku) [reply Assets/Scripts/UI/MainMenuController.cs:23]: Should probably have a getDefaultMenu method instead of making it public in that case

File: Assets/Scripts/Augment/GunFactory.cs
Comment (Fueredoriku) [inline Assets/Scripts/Augment/GunFactory.cs:138]: Maybe add a // TODO: Implement save/get instead og norwegian comments

**Review by Fueredoriku — CHANGES_REQUESTED** [review]
I think this is looking nice already!

I added a comment and some responses that needs to be looked at
Also one thing needs to be addressed before this is approved:
the Main menu now defaults to starting in the gallery instead of the landing page, this should be reverted before merge:
image

File: Assets/Scripts/UI/MainMenuController.cs
Comment (toberge) [reply Assets/Scripts/UI/MainMenuController.cs:23]: ...should probably have a **property** with private backing field, please avoid explicit get methods for this smh

File: Assets/Scripts/Augment/GunFactory.cs
Comment (toberge) [reply Assets/Scripts/Augment/GunFactory.cs]: oh
oh hold on
I didn't see the extra parameters
um
yeah okay though that's some functionality that is nice to have
maybe do an overload but write it so that it calls the other InstantiateGun instead of duplicating the code :/

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [reply Assets/Scripts/UI/GalleryMenu.cs:198]: Um, I was talking about the `sInfo` abbreviation

File: Assets/Editor/WeaponDrawer.cs
Comment (toberge) [reply Assets/Editor/WeaponDrawer.cs:9]: As stated in a PM, this could be removed since Unity does have editor GUI for structs by default :upside_down_face:

File: Assets/Scripts/Augment/GunFactory.cs
Comment (TArchetype) [reply Assets/Scripts/Augment/GunFactory.cs]: According to @SondreHus, the overload was unnecessary. So I'm opting to keep it as it is now.

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (TArchetype) [reply Assets/Scripts/UI/GalleryMenu.cs:198]: You got to be more specific man! But I've renamed the variable name properly now.

File: Assets/Scripts/Augment/GunFactory.cs
Comment (TArchetype) [reply Assets/Scripts/Augment/GunFactory.cs]: Comments and pseudocode has been removed.

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [reply Assets/Scripts/UI/GalleryMenu.cs]: Sorry, I thought it'd be evident from the line I marked 🙃

**Comment by toberge** [comment]
> I think this is looking nice already!
> 
> I added a comment and some responses that needs to be looked at Also one thing needs to be addressed before this is approved: the Main menu now defaults to starting in the gallery instead of the landing page, this should be reverted before merge: image

This appears to be fixed now

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [inline Assets/Scripts/UI/GalleryMenu.cs:92]: `DestroyImmediate` seems dangerous according to [Unity's documentation](https://docs.unity3d.com/ScriptReference/Object.DestroyImmediate.html), could we get away with using `Destroy` here or is there a reason you use the _immediate_ version?

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [inline Assets/Scripts/UI/GalleryMenu.cs:127]: Perhaps we should turn this into a method, since you're using it twice here and it's a little convoluted
(also, you're setting the layer in both loops, is that necessary?)

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [inline Assets/Scripts/UI/GalleryMenu.cs:144]:
```cs
}
else
{
```

please use autoformatting :)))))

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (toberge) [inline Assets/Scripts/UI/GalleryMenu.cs:209]: You could do this extension-less addition _before_ the extension loop, having the loop start at `-1` and then not using the loop variable is a tad too confusing.

**Review by toberge — CHANGES_REQUESTED** [review]
The gallery behaves nicely now, and no longer takes over the main menu. Visual adjustments and statistics functionality can be left for later. Please have a look at the comments I left. When those are resolved, this PR should be ripe for merging.

File: Assets/Scripts/UI/FlexibleGridLayout.cs
Comment (TArchetype) [reply Assets/Scripts/UI/FlexibleGridLayout.cs:52]: Somehow this seems to be working now?

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (TArchetype) [reply Assets/Scripts/UI/GalleryMenu.cs:92]: changed

File: Assets/Scripts/UI/GalleryMenu.cs
Comment (TArchetype) [reply Assets/Scripts/UI/GalleryMenu.cs:144]: ...

**Review by toberge — APPROVED** [review]
yeah it's good now

**Review by Fueredoriku — APPROVED** [review]
Diff looks good!

I can't test atm due to porting version, but I'll approve based on the code

**Pull request merged**
